### PR TITLE
Task/api auth

### PR DESCRIPTION
**Pull request by alszczep**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,11 +2,25 @@
 
 This project was generated using [Nx](https://nx.dev).
 
-## Build
+### [pdf-renderer](https://github.com/leancodepl/pdf-generator/tree/main/packages/pdf-renderer)
```
Comment (konowrockis) [inline README.md:5]: :thinking: wondering if the url here shouldn't be relative. So if for example you're looking at readme on one branch you don't get redirected always to main.

File: package.json [hunk 2]
```diff
@@ -5,12 +5,19 @@
...
     "@nestjs/core": "^8.0.0",
+    "@nestjs/passport": "^8.1.0",
     "@nestjs/platform-express": "^8.0.0",
     "core-js": "^3.6.5",
     "date-fns": "^2.28.0",
-    "puppeteer": ">=13.0.0",
+    "jwks-rsa": "^2.0.5",
+    "passport": "^0.5.2",
+    "passport-jwt": "^4.0.0",
+    "puppeteer": "^9.0.0",
```
Comment (konowrockis) [inline package.json:20]: Why it got downgraded from `13`?

File: packages/api-proxy/README.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+based api.
+
+## Usage
+
+### register
+
+```ts
+import { VerifyOptions } from "jsonwebtoken";
+
+static register({ isGlobal = false, jwtStrategyConfig }: ApiProxySyncConfiguration): DynamicModule
```
Comment (konowrockis) [inline packages/api-proxy/README.md:13]: `isGlobal = false` is not exactly true anymore.
`static` in this context looks kinda weird. IMO it'd be better if you'd do `ApiProxy.register({ ...`

File: packages/api-proxy/README.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+You can specify if you want to register a module globally.
+
+##### jwksUri
+
+Uri of your auth jwks.
+
+##### jsonWebTokenOptions
+
+`VerifyOptions` is a type imported from `jsonwebtoken` package.
```
Comment (konowrockis) [inline packages/api-proxy/README.md:36]: Here you can either show the `VerifyOptions` type or link to the originical documentation if there's any.

File: packages/api-proxy/README.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+Remember to add those classes as providers of your module.
+
+```ts
+@Module({
+    imports: [ApiProxyModule.register(apiAndAuthConfig), PassportModule.register({ defaultStrategy: "jwt" })],
+    providers: [CqrsClient1, Query1ComponentService],
+})
+export class AppModule {}
+```
```
Comment (konowrockis) [inline packages/api-proxy/README.md:156]: There's no information about the need of using `@UseJwtGuard()` and that's actually crucial to the process.

File: packages/api-proxy/src/lib/api-proxy.module.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+export interface ApiProxyAsyncConfiguration extends Pick<ModuleMetadata, "imports"> {
+    useExisting?: Type<ApiProxyConfigurationFactory>;
+    useClass?: Type<ApiProxyConfigurationFactory>;
+    useFactory?: (...args: any[]) => Promise<ApiProxyConfiguration> | ApiProxyConfiguration;
+    inject?: any[];
+    extraProviders?: Provider[];
+    isGlobal?: boolean;
+}
+
+@UseJwtGuard()
```
Comment (konowrockis) [inline packages/api-proxy/src/lib/api-proxy.module.ts:31]: ?

File: packages/api-proxy/src/lib/api.service.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import { map } from "rxjs/operators";
+import { CqrsClient } from "./cqrsClient";
+
+export type EndpointGetter = (type: string) => string;
+
+@Injectable()
+export class Api implements CqrsClient {
+    constructor(
+        private httpService: HttpService,
+        private request: Request,
```
Comment (konowrockis) [inline packages/api-proxy/src/lib/api.service.ts:15]: actually all of these can be `readonly`

File: packages/api-proxy/src/lib/cqrsClientFactory.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { Api, EndpointGetter } from "@leancodepl/api-proxy";
+import { HttpService } from "@nestjs/axios";
+import { Inject, Injectable } from "@nestjs/common";
+import { REQUEST } from "@nestjs/core";
+import type { Request } from "express";
+
+@Injectable()
+export class CqrsClientFactory {
+    constructor(private httpService: HttpService, @Inject(REQUEST) private request: Request) {}
```
Comment (konowrockis) [inline packages/api-proxy/src/lib/cqrsClientFactory.ts:9]: `readonly`

File: packages/sample-project/src/app/app.module.ts [hunk 9]
```diff
@@ -1,14 +1,41 @@
...
+import { CqrsClient1 } from "./components-services/CqrsClient1";
 import { InvoiceTemplateService } from "./components-services/invoiceTemplate.service";
+import { Query1ComponentService } from "./components-services/query1Component.service";
 import { SampleComponentService } from "./components-services/sampleComponent.service";
 import { fontsConfiguration } from "./fonts";
+import { QueryController } from "./query.controller";
+
+const apiAndAuthConfig: ApiProxyConfiguration = {
+    jwtStrategyConfig: {
+        jwksUri: "https://api.prospector.test.lncd.pl/auth/.well-known/openid-configuration/jwks",
```
Comment (konowrockis) [inline packages/sample-project/src/app/app.module.ts:16]: Oh, I think we were discussing of injecting that from env and that'd be good example for the sample app as well.

File: packages/sample-project/src/app/components-services/CqrsClient1.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+@Injectable()
+export class CqrsClient1 {
+    client;
+    constructor(cqrsClientFactory: CqrsClientFactory) {
+        this.client = Client(cqrsClientFactory.create(CqrsClient1.getApiEndpoint));
+    }
+
+    static getApiEndpoint(type: string) {
+        return `http://localhost:3333/api/${type}`;
```
Comment (konowrockis) [inline packages/sample-project/src/app/components-services/CqrsClient1.ts:13]: This also should come from the env.

File: workspace.json [hunk 11]
```diff
@@ -1,6 +1,40 @@
 {
   "version": 2,
   "projects": {
+    "api-proxy": {
```
Comment (konowrockis) [inline workspace.json:4]: Wait, why did you switch to workspace.json and not per package config?

**Review by konowrockis — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (alszczep) [reply package.json:20]: because versions >= 10 use this https://github.com/advisories/GHSA-r683-j2x4-v87g

**Review by konowrockis — APPROVED** [review]

**Pull request merged**